### ISSUE TITLE
Does shmap work with jax.lax.scan inside the mapped function?

### ISSUE THREAD

**Issue by minqi** [google/jax#15996]
### Description

I am attempting to perform a sharded_map of a function that has an inner jax.lax.scan (via `fori_loop`). When I run the shmapped function, the result is an error complaining that there is `No replication rule for scan`.

For further context, I am attempting to switch from `pmap` to `shmap`, as I would like to perform an outer `jit` on a function that calls the pmapped function internally. Inside this function, I call pmap over an axis named `'devices'` and the output of the pmapped function is further processed by other methods (some of which will be jitted as a result of the outer jit). However, the result is an error that states the `'devices'` axes is not available. My understanding is that this is due to pjit and pmap using fundamentally incompatible axes abstractions.

### What jax/jaxlib version are you using?

4.8

### Which accelerator(s) are you using?

GPU

### NVIDIA GPU info

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.141.03   Driver Version: 470.141.03   CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Quadro GP100        Off  | 00000000:81:00.0 Off |                    0 |
| 26%   38C    P0    30W / 235W |      0MiB / 16278MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Quadro GP100        Off  | 00000000:82:00.0 Off |                    0 |
| 26%   36C    P0    30W / 235W |      0MiB / 16278MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

**Comment by mattjj** [comment]
Thanks for the question! Sorry for not seeing it until now.

Actually this error should have been fixed in #15481. Can you try updating your JAX version past 0.4.10 and see if the error persists?

By the way, this error is a pretty innocuous one; it's just saying that a safety feature isn't (well, wasn't) available for `scan`. But we can disable that safety feature, and hence avoid this kind of NotImplementedError, by passing `check_rep=False` to the `shard_map` call. (We should probably add that information to the error message itself.)

> For further context, I am attempting to switch from pmap to shmap, as I would like to perform an outer jit on a function that calls the pmapped function internally. Inside this function, I call pmap over an axis named 'devices' and the output of the pmapped function is further processed by other methods (some of which will be jitted as a result of the outer jit). However, the result is an error that states the 'devices' axes is not available. My understanding is that this is due to pjit and pmap using fundamentally incompatible axes abstractions.

I'd have to see a repro to know for sure, but yes in general `pmap` and `pjit`/`jit` are incompatible (at least as far as _efficiency_ is concerned), whereas `shard_map` + `pjit`/`jit` are happy with each other!

Since I'm 95+% sure that #15481 fixed this issue, I'm going to speculatively close it as resolved. But if I'm mistaken please let me know (and reopen it if you can)!
